contracts/Rewards.sol

```diff
@@ -0,0 +1,199 @@
+pragma solidity 0.8.6;
+
+/// @title Rewards
+/// @notice Rewards are allocated proportionally to operators
+/// present in the pool at payout based on their weight in the pool.
+///
+/// To facilitate this, we use a global accumulator value
+/// to track the total rewards one unit of weight would've earned
+/// since the creation of the pool.
+///
+/// Whenever a reward is paid, the accumulator is increased
+/// by the size of the reward divided by the total weight
+/// of all eligible operators in the pool.
+///
+/// Each operator has an individual accumulator value,
+/// set to equal the global accumulator when the operator joins the pool.
+/// This accumulator reflects the amount of rewards
+/// that have already been accounted for with that operator.
+///
+/// Whenever an operator's weight in the pool changes,
+/// we can update the amount of rewards the operator has earned
+/// by subtracting the operator's accumulator from the global accumulator.
+/// This gives us the amount of rewards one unit of weight has earned
+/// since the last time the operator's rewards have been updated.
+/// Then we multiply that by the operator's previous (pre-change) weight
+/// to determine how much rewards in total the operator has earned,
+/// and add this to the operator's earned rewards.
+/// Finally, we set the operator's accumulator to the global accumulator value.
+contract Rewards {
+  struct OperatorRewards {
+    // The state of the global accumulator
+    // when the operator's rewards were last updated
+    uint96 accumulated;
+    // The amount of rewards collected by the operator after the latest update.
+    // The amount the operator could withdraw may equal `available`
+    // or it may be greater, if more rewards have been paid in since then.
+    uint96 available;
+    // If nonzero, the operator is ineligible for rewards
+    // and may only re-enable rewards after the specified timestamp.
+    // XXX: unsigned 32-bit integer unix seconds, will break around 2106
+    uint32 ineligibleUntil;
+    // Locally cached weight of the operator,
+    // used to reduce the cost of setting operators ineligible.
+    uint32 weight;
+  }
+
+  // The global accumulator of how much rewards
+  // a hypothetical operator of weight 1 would have earned
+  // since the creation of the pool.
+  uint96 internal globalRewardAccumulator;
+  // If the amount of reward tokens paid in
+  // does not divide cleanly by pool weight,
+  // the difference is recorded as rounding dust
+  // and added to the next reward.
+  uint96 internal rewardRoundingDust;
+
+  // The amount of rewards that would've been earned by ineligible operators
+  // had they not been ineligible.
+  uint96 internal ineligibleEarnedRewards;
```

Comment:
I am thinking it could make sense to make it public in case the DAO wants to check if it makes sense to call the withdrawal or wait.

Reply:
:+1: